test/core/event_engine/fuzzing_event_engine/fuzzing_event_engine.h

```diff
@@ -48,62 +57,196 @@ class FuzzingEventEngine : public EventEngine {
+  // each test.
+  void UnsetGlobalHooks() ABSL_LOCKS_EXCLUDED(mu_);
 
  private:
+  // One pending task to be run.
   struct Task {
     Task(intptr_t id, absl::AnyInvocable<void()> closure)
         : id(id), closure(std::move(closure)) {}
     intptr_t id;
     absl::AnyInvocable<void()> closure;
   };
 
+  // Per listener information.
+  // We keep a shared_ptr to this, one reference held by the FuzzingListener
+  // Listener implementation, and one reference in the event engine state, so it
+  // may be iterated through and inspected - principally to discover the ports
+  // on which this listener is listening.
+  struct ListenerInfo {
+    ListenerInfo(
+        Listener::AcceptCallback on_accept,
+        absl::AnyInvocable<void(absl::Status)> on_shutdown,
+        std::unique_ptr<MemoryAllocatorFactory> memory_allocator_factory)
+        : on_accept(std::move(on_accept)),
+          on_shutdown(std::move(on_shutdown)),
+          memory_allocator_factory(std::move(memory_allocator_factory)),
+          started(false) {}
+    ~ListenerInfo() ABSL_LOCKS_EXCLUDED(mu_);
+    // The callback to invoke when a new connection is accepted.
+    Listener::AcceptCallback on_accept;
+    // The callback to invoke when the listener is shut down.
+    absl::AnyInvocable<void(absl::Status)> on_shutdown;
+    // The memory allocator factory to use for this listener.
+    const std::unique_ptr<MemoryAllocatorFactory> memory_allocator_factory;
+    // The ports on which this listener is listening.
+    std::vector<int> ports ABSL_GUARDED_BY(mu_);
+    // Has start been called on the listener?
+    // Used to emulate the Bind/Start semantics demanded by the API.
+    bool started ABSL_GUARDED_BY(mu_);
+    // The status to return via on_shutdown.
+    absl::Status shutdown_status ABSL_GUARDED_BY(mu_) = absl::OkStatus();
+  };
+
+  // Implementation of Listener.
+  class FuzzingListener final : public Listener {
+   public:
+    explicit FuzzingListener(std::shared_ptr<ListenerInfo> info)
+        : info_(std::move(info)) {}
+    ~FuzzingListener() override;
+    absl::StatusOr<int> Bind(const ResolvedAddress& addr) override;
+    absl::Status Start() override;
+
+   private:
+    std::shared_ptr<ListenerInfo> info_;
+  };
+
+  // One read that's outstanding.
+  struct PendingRead {
+    // Callback to invoke when the read completes.
+    absl::AnyInvocable<void(absl::Status)> on_read;
+    // The buffer to read into.
+    SliceBuffer* buffer;
+  };
+
+  // The join between two Endpoint instances.
+  struct EndpointMiddle {
+    EndpointMiddle(int listener_port, int client_port)
+        ABSL_EXCLUSIVE_LOCKS_REQUIRED(mu_);
+    // Address of each side of the endpoint.
+    const ResolvedAddress addrs[2];
+    // Is the endpoint closed?
+    bool closed ABSL_GUARDED_BY(mu_) = false;
+    // Bytes written into each endpoint and awaiting a read.
+    std::vector<uint8_t> pending[2] ABSL_GUARDED_BY(mu_){
```

Comment:
These are default-initialized, right? I think the assignment is redundant.

Reply:
ack